### ISSUE TITLE
In order to mitigate sybil attacks, peers should detect unusual request patterns

### ISSUE THREAD

**Issue by sanity** [freenet/freenet-core#686]
Peers should look for unusual statistical patterns in requests coming from their neighbors. For example, a bunch of new contracts being added all with unusually similar locations might be indicative of a Sybil attack. Peers with very unusual such patterns should be disconnected.

**Comment by Destroyinator69420** [comment]
I agree, there should be a strict definition of how a locutus node should behave. If a locutus node does not fit that pattern, then the connection should be dropped. If it does not behave exactly like the reference implementation, (the locutus node in this repository) then it should be shunned from the network. I also think that the symmetric encryption locutus uses should not place trust in a single algorithm. I would recommend AES-256(Kuznyechik-256(Kalyna-512(Message))). We shoudl use these algorithms in XTS mode. It would help the censorship resistance of Locutus if all messages sent through the network are of one standard size in the ciphertext, but from an outside observer they are compressed with ZSTD and padded by no more than 10%.

**Comment by ple1n** [comment]
Again, that's not really how censorship resistance works. You get easily detected if your packets are statistically uniform and 'featureless', and/or of same size. This is proven by https://gfw.report research and such.

**Comment by sanity** [comment]
🤖 **Relabeled for consistency**

Applied labels:
- `T-feature`
- `P-high`
- `E-hard`
- `A-networking`

**Reasoning:** This is a request for new functionality (detecting statistical request patterns to mitigate Sybil attacks), which is security-relevant so deserves high priority. Implementing reliable detection and safe disconnection behavior requires non-trivial heuristics, protocol consideration and careful testing, so effort is estimated as hard. The change primarily affects peer/network behaviour.

Previous labels: `T-feature`

**Comment by sanity** [comment]
# Issue Priority Assessment

Generated: 2026-02-22T14:53:11.166481357+00:00
Issue hash: `6bee96de`

## Value Assessment

This issue touches every running node because the `Node` type (the main event loop that orchestrates peer connections, operations, and client interactions) is the central code path for Freenet’s distributed processing, so any neighbor whose requests can overwhelm or skew those queues affects the entire peer population; the ability to detect and sever such neighbors safely thus matters to every user’s experience, not just the handful that might take the hit directly. When a peer is flagged as suspicious, the only sanctioned way to sever the session is `NetworkBridge::drop_connection` (`crates/core/src/node/network_bridge.rs:31`), so new heuristics would need to feed into that hook to avoid letting a Sybil-style flood simply keep circling the node’s request queues. Without such a mechanism, the node can still receive requests for PUT/GET/UPDATE operations, but it has no statistical guardrail and can be forced to spend cycles lining up identical contract operations; the resulting latency and dropped client requests would degrade service for everyone sharing that route.

Strategically, the transport security documentation lays down the project’s security goals—peer identity, confidentiality, authenticity, and simplicity at the core of the threat model (`docs/architecture/transport/security.md:5`)—and explicitly marks traffic-pattern analysis and related Sybil-style behaviors as out of scope (`docs/architecture/transport/security.md:92`). That means the project acknowledges the threat but currently accepts the exposure; building neighbor-request pattern detection would bring a much-needed defense that slots directly into the existing threat model instead of changing directions, advancing the core goal of keeping the overlay resilient without altering the cryptographic primitives already in place.

In terms of blocking other work, almost every higher-level feature around routing, deduplication, and contract delivery assumes that peers behave roughly within expected statistical bounds (the `Node` request router is already fighting to deduplicate GET/PUT/UPDATE requests before things hit the operation pipelines, `crates/core/src/node/request_router.rs:1`). If Sybil peers can inflate certain request types or locations, those deduplication caches and the operations they spawn will stay bogged down, making it hard to harden other areas like protocol upgrades or richer subscription semantics without first stabilizing the peer population. Implementing this issue therefore clears a necessary pathway for future enhancements that rely on trustworthy neighbors and sane traffic.

As for urgency, while there is no fixed calendar deadline, the issue was tagged as security-relevant because the current infrastructure explicitly gives Sybil attackers room to drive request storms without triggering disconnection, and such attacks can be triggered at will once the adversary has enough nodes. Given that traffic pattern abuse has been noted as unmitigated, any delay just extends the window during which a malfunction or adversary can wedge the node’s operation queues, so prioritizing heuristics that score neighbor request similarity should happen as soon as bandwidth is available.

## Cost Assessment

Putting the detection inside the existing event loop means touching the message path around `handle_inbound_message`/`process_message` in `crates/core/src/node/network_bridge/p2p_protoc.rs:2068-2143` so you can see every inbound `NetMessage` before it’s handed to operations, and you’ll need to consult the same module’s `ConnectionEntry` bookkeeping (`…/p2p_protoc.rs:338-397`) to attribute those messages to the right neighbor identity/`TransportPublicKey`. The `TopologyManager`/`ConnectionManager` stack already pushes requests out and knows about every connected peer, so whatever heuristic you build will also have to tie into `crates/core/src/topology/mod.rs:82-138` and whatever part of the ring needs to react to the new “drop this neighbor” action, which today flows through `NodeEvent::DropConnection` in `crates/core/src/message.rs:588-662` and triggers `P2pBridge::handle_orphaned_transactions` at `…/p2p_protoc.rs:148-190`. All of that means at least the `node/network_bridge` loop, the ring connection tracker, and the event-routing glue need updates.

The statistical work itself is non-trivial because there’s no existing inbound tracker—only the outbound counterpart (`crates/core/src/topology/outbound_request_counter.rs:4-41`) and the request-density helper that scores outgoing targets (`crates/core/src/topology/request_density_tracker.rs:7-88`). You would need to mirror that sliding-window, per-neighbor bookkeeping while using `MessageStats::requested_location` in `crates/core/src/message.rs:269-273` to see whether an incoming stream of requests is clustering around a suspicious location. Placing that logic in the event loop is delicate because `process_message` immediately offloads work via `GlobalExecutor::spawn` (`…/p2p_protoc.rs:2131-2143`), so the tracker must be thread-safe and able to survive the asynchronous detours that currently handle each `NetMessage`, or else you risk losing samples or deadlocking.

Any automated disconnection at this layer raises safety concerns because a false positive walks straight through the same `NodeEvent::DropConnection`/`P2pBridge::drop_connection` path (`message.rs:588-662` and `p2p_protoc.rs:148-190`) and may orphan operations that were mid-flight. The connection manager and `OpManager` expect clean failure handling, so the detector must also emit the standard orphaned-transaction recovery flow (see the logging and retry logic in `handle_orphaned_transactions`). Tuning thresholds will be risky without observability, so configuration/telemetry hooks will probably need additions near where topology currently records densities and request counts.

Testing will also grow overnight: you’ll need unit coverage for the new sliding-window tracker (mirror the style of the existing `request_density_tracker` tests around `crates/core/src/topology/request_density_tracker.rs:330-418`) and integration/simulation hooks that feed synthetic neighbors stream patterns to the `p2p_protoc` event loop. Those tests must prove that benign bursts don’t trigger the detector, that false positives roll into the same cleanup paths already exercised by `handle_orphaned_transactions`, and that there’s no regression in the threading-sensitive `process_message`/`GlobalExecutor` plumbing.

---
*Generated by issue-prioritizer*

**Comment by sanity** [comment]
🤖 **Automated ranking by Torvan.** Not a maintainer's assessment of this issue.

| Measure | Score |
|---|---:|
| Value | 0.432 |
| Cost | 0.931 |
| Priority | 435 of 593 |

Value and cost run 0 to 1 across the 593 open issues in the freenet repositories: value 1.00 is the most valuable, cost 0.00 the cheapest. Issues are compared against each other rather than scored in isolation. The queue alternates between the most valuable work and the best value for the effort, so neither expensive-but-important nor cheap-and-quick work crowds the other out.

Ranked 2026-08-31, and liable to change. Adding information here, whether that is scope, a reproduction, or a correction to this assessment, causes the issue to be reassessed and this comment updated.
